src/core/shape_inference/include/multiclass_nms_shape_inference.hpp

```diff
@@ -0,0 +1,203 @@
+            op,
+            scores_ps.rank().is_static() && (scores_ps.rank().get_length() == 3 || scores_ps.rank().get_length() == 2),
+            "Expected a 2D or 3D tensor for the 'scores' input. Got: ",
+            scores_ps);
+    }
+
+    if (op->get_input_size() == 3) {
+        const auto& roisnum_ps = input_shapes[2];
+        NODE_VALIDATION_CHECK(op,
+                              roisnum_ps.rank().is_static() && roisnum_ps.rank().get_length() == 1,
+                              "Expected a 1D tensor for the 'roisnum' input. Got: ",
+                              roisnum_ps);
+    }
+
+    // validate compatibility of input shapes
+    if (scores_ps.rank().is_static() && scores_ps.rank().get_length() == 3) {  // if scores shape (N, C, M)
+        const auto num_batches_boxes = boxes_ps[0];
+        const auto num_batches_scores = scores_ps[0];
+
+        NODE_VALIDATION_CHECK(op,
+                              num_batches_boxes.same_scheme(num_batches_scores),
+                              "The first dimension of both 'boxes' and 'scores' must match. Boxes: ",
+                              num_batches_boxes,
+                              "; Scores: ",
+                              num_batches_scores);
+
+        const auto num_boxes_boxes = boxes_ps[1];
+        const auto num_boxes_scores = scores_ps[2];
+        NODE_VALIDATION_CHECK(op,
+                              num_boxes_boxes.same_scheme(num_boxes_scores),
+                              "'boxes' and 'scores' input shapes must match at the second and third "
+                              "dimension respectively. Boxes: ",
+                              num_boxes_boxes,
+                              "; Scores: ",
+                              num_boxes_scores);
+    }
+
+    if (scores_ps.rank().is_static() && scores_ps.rank().get_length() == 2) {  // if scores shape (C, M)
+        const auto num_classes_boxes = boxes_ps[0];
+        const auto num_classes_scores = scores_ps[0];
+        NODE_VALIDATION_CHECK(op,
+                              num_classes_boxes.same_scheme(num_classes_scores),
+                              "'boxes' and 'scores' input shapes must match. Boxes: ",
+                              num_classes_boxes,
+                              "; Scores: ",
+                              num_classes_scores);
+
+        const auto num_boxes_boxes = boxes_ps[1];
+        const auto num_boxes_scores = scores_ps[1];
+        NODE_VALIDATION_CHECK(op,
+                              num_boxes_boxes.same_scheme(num_boxes_scores),
+                              "'boxes' and 'scores' input shapes must match. Boxes: ",
+                              num_boxes_boxes,
+                              "; Scores: ",
+                              num_boxes_scores);
+
+        NODE_VALIDATION_CHECK(op,
+                              op->get_input_size() == 3,
+                              "Expected the 'roisnum' input when the input 'scores' is a 2D tensor.");
```

Comment:
Just a nitpick, you can move it to the top of if-scope. As there is no need to extract any info from boxes/scores if this isn't 3-inputs case after all.